### ISSUE TITLE
API Proposal: RuntimeFeature.HasJit and RuntimeFeature.CanEmitCode

### ISSUE THREAD

**Issue by morganbr** [dotnet/corefx#29258]
# Rationale
The most efficient patterns for reflection depend on the current runtime. On .NET Core, ```System.Reflection.
Emit``` or LINQ Expressions are usually most efficient because the JIT compiles them at runtime. However, for .NET Native, CoreRT, (and I believe Mono AoT), ```System.Reflection.Emit``` isn't available and LINQ Expressions are interpreted. Mono has progress on an interpreter that could make ```System.Reflection.Emit``` work, but it would still likely be less efficient than using reflection APIs directly. Library developers occasionally ask for a way to find out what will both work and be efficient on the current runtime. 

This has come up in #10470 and https://github.com/aspnet/Home/issues/3079.

# Proposed API
Add ```HasJit``` and ```CanEmitIL``` properties to the existing ```System.Runtime.CompilerServices.RuntimeFeature``` class:
```
public class RuntimeFeature
{
    public static bool HasJit { get; }
    public static bool CanEmitIL { get; }
}
```

```HasJit``` is true if the current runtime can emit optimized native code at runtime. ```CanEmitIL``` is true if System.Reflection.Emit works at all, whether it's implemented by a JIT or an interpreter. 

# Example Usage
Serializers or dependency injection frameworks often have methods that construct an object with some knowledge of its signature. With this API, it could be optimized to look something like this:
```
T CreateInstance<T>(int ctorParameter)
{
    if (RuntimeFeature.HasJit)
    {
        // Since this runtime has a JIT, use refemit to construct a specialized method
        if(cachedFactoryMethodDelegate == null)
        {
            DynamicMethod factoryMethod = new DynamicMethod(...);
            // Emit some IL to instantiate T
            cachedFactoryMethodDelegate = (Func<int, T>)factoryMethod.CreateDelegate(typeof(Func<int, T>));
         }
         return cachedFactoryMethodDelegate(ctorParameter);
    }
    else
    {
        // No JIT, just use reflection
        ConstructorInfo ctorInfo = typeof(T).GetConstructor(new Type[] { typeof(int)});
        return (T)ctorInfo.Invoke(ctorParameter);
     }
}

**Comment by jkotas** [comment]
We have duplicate issues of this:

https://github.com/dotnet/corefx/issues/13535
https://github.com/dotnet/corefx/issues/25530

The preferred design for API availability checks in earlier discussions was to just add static `IsSupported` property to the API, like `AssemblyBuilder.IsSupported` or `DynamicMethod.IsSupported`. This pattern  scalles well to 3rd party APIs.

`RuntimeFeature.HasJit` makes sense to me.

**Comment by morganbr** [comment]
The IsSupported pattern would make this proposal look like this (with an IsSupported method for all entrypoint classes for reflection emit). What do folks think?

```
namespace System.Runtime.CompilerServices
{
    public class RuntimeFeature
    {
        public static bool HasJit { get; }
    }
}

namespace System.Reflection.Emit
{
    public class AssemblyBuilder
    {
        public static bool IsSupported { get; }
    }

    public class DynamicMethod
    {
        public static bool IsSupported { get; }
    }

    public class CustomAttributeBuilder
    {
        public static bool IsSupported { get; }
    }
}
```

**Comment by jkotas** [comment]
LGTM. 

I am not sure whether we need `CustomAttributeBuilder.IsSupported`. Can `CustomAttributeBuilder` be ever meaningfully used without `AssemblyBuilder`?

**Comment by morganbr** [comment]
We could probably do without ```CustomAttributeBuilder.IsSupported```. On closer inspection, the only thing it does by itself is build a binary representation of the custom attribute. While it's not in CoreRT right now, it could be as long as we drop the internal methods used by other parts of reflection emit.

After a bit of digging, we could potentially need more IsSupported APIs to cover downstream usage refemit. For example, Regex and System.Composition have APIs that depend on it. Would we want some sort of support check on those as well?

**Comment by morganbr** [comment]
CC @davidfowl, who could hopefully use this in ASP.NET.

**Comment by jkotas** [comment]
Yes, I think having a property like `bool Regex.IsCompiledSupported` would make sense.

**Comment by davidfowl** [comment]
We should walk through the flow chart on what this would look like for our DI container. Today we have a couple of different modes (https://github.com/aspnet/DependencyInjection/blob/943ffa1187399ad4c2af0d23641af9804eacd4e6/src/DI/ServiceProviderMode.cs#L6):
- Dynamic - Use reflection the first 2 times somebody asks to resolve a service, then switch then compile it in the background after that using the most efficient method (Ref Emit, or Compiled expressions)
- Runtime - Always use reflection when resolving services
- Expressions - Compile the expression tree before resolving service
- ILEmit - Compile a dynamic method before resolving service

Here's what I'm thinking the flow control would be:

// Set the default mode to runtime if there's no JIT
https://github.com/aspnet/DependencyInjection/blob/06e2de235ce5b27b425e823d9dcbd045811ba48e/src/DI/ServiceProviderOptions.cs#L21

```C#
Mode = RuntimeFeature.HasJit ? ServiceProviderMode.Dynamic : ServiceProviderMode.Runtime;
```

https://github.com/aspnet/DependencyInjection/blob/06e2de235ce5b27b425e823d9dcbd045811ba48e/src/DI/ServiceLookup/ServiceProviderEngine.cs#L23
```
// This name should be CompiledBuilder
ExpressionBuilder = DynamicMethod.IsSupported ? new ILEmitResolver(...) : new CallSiteExpressionBuilder(...);
```

/cc @pakrym for DI 
/cc @JamesNK (he originally asked for this) for JSON.NET

**Comment by morganbr** [comment]
@davidfowl, now that I think about it a bit more, is ```DynamicMethod.IsSupported``` useful or should we only have ```RuntimeFeatures.HasJit```? In other words if refemit is interpreted, would you ever want to use it?

**Comment by davidfowl** [comment]
> @davidfowl, now that I think about it a bit more, is DynamicMethod.IsSupported useful or should we only have RuntimeFeatures.HasJit? In other words if refemit is interpreted, would you ever want to use it?

No we'd never want to use it if it was interpreted in this scenario unless it was faster than reflection.

**Comment by morganbr** [comment]
Ok, let's just go with ```RuntimeFeatures.HasJit``` until/unless we see a useful scenario for using RefEmit with no JIT.

**Comment by morganbr** [comment]
I incorrectly thought ```RuntimeFeatures``` was public. Would ```System.Runtime.InteropServices.RuntimeInformation``` be the right public class @stephentoub?

**Comment by jkotas** [comment]
@morganbr `System.Runtime.CompilerServices.RuntimeFeature` is public: https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimefeature?view=netcore-2.0

**Comment by morganbr** [comment]
Ah, right you are. I just didn't find it in the contract sources due to magic in how System.Runtime is built. In that case, I'll go ahead and mark this ready for review.

**Comment by morganbr** [comment]
@terrajobst , the point of HasJit is that some features, like compiled LINQ expressions may work on AoT, but they're slow. That might eventually be true of DynamicMethod/Assemblybuilder as well. In those cases, IsSupported would be true, but a library like ASP.NET or Json.NET would want to find out whether they make code faster than reflection (because there's a JIT) or slower (because they're interpreted). 

I'm open to putting HasJit on a different class.

**Comment by JamesNK** [comment]
> This is a dynamic feature discovery. It seems this should be DynamicMethod.IsSupported and AssemblyBuilder.IsSupported.

I wanted to use a flag like this to test whether `System.Linq.Expressions` would get compiled or interpreted after calling `Compile()`. This flag isn't just specific to `DynamicMethod` so I don't think it should be placed on `DynamicMethod.`.

**Comment by jkotas** [comment]
> We still have to figure out how either of these two can be used in the context of .NET Standard 

The proposed API is for inclusion in the .NET Core platform. If somebody wants to build a package with API that does best effort approximation of this API for existing .NET Standard 2.0 platforms, that's fine but it should be different assembly and package name that does not unify with the .NET Core platform one.

**Comment by morganbr** [comment]
I'd rather have HasJit in .NET Standard, particularly since it's relevant in UAP. AssemblyBuilder/DynamicMethod are different and not part of this proposal.

**Comment by jkotas** [comment]
If we agree on a reasonable API, I am sure it will be included in .NET Standard 3.0. .NET Standard 2.0 is done. There is no way to add new APIs to it.

**Comment by morganbr** [comment]
Agreed. We'll just need to bring it to UAP as well as .NET Core (and perhaps some Xamarin flavors too).

**Comment by davidfowl** [comment]
So is the api ready for review?

**Comment by terrajobst** [comment]
This is related: https://github.com/dotnet/standard/issues/832

I believe this feature needs a more comprehensive design. I'll follow-up.

**Comment by terrajobst** [comment]
I've updated the proposal based on our meeting notes. I hope @morganbr doesn't mind it shows up as his :-)

**Comment by morganbr** [comment]
Heh, the update is fine (though I don't think there's actually a penalty for using compiled regex, it just uses the state machine).

**Comment by terrajobst** [comment]
The string constants are meant for static discovery of runtime features. I believe these capabilities are inherently dynamic, so making them string constants feels wrong (note that the presence of the string constants presupposes the runtime must runtime `true` from `IsSupported(string)`.

**Comment by pakrym** [comment]
Are we going to backport the implementation of `RuntimeFeature.IsSupported` to 2.1/2.2?

If we don't, any app that's targeting netcoreapp lower then 3.0 and using `RuntimeFeature.IsSupported("IsDynamicCodeCompiled")` would be told that compilation is not supported.

**Comment by darkl** [comment]
Hi @terrajobst,
Can I use `RuntimeFeature` to detect whether `DispatchProxy` is supported on the current platform?

Thanks and happy holidays
Elad

**Comment by darkl** [comment]
Anyone else who knows the answer is also welcome to reply :)

**Comment by jkotas** [comment]
@darkl We are not monitoring old closed issues, and your question is somewhat unrelated to what this issue about. If you would like to get answer to your question, open a new question issue in https://github.com/dotnet/runtime
